File: cgi-bin/php/download.php
<?php

require_once("utils.php");

$path = getenv('ROOT_PATH') . ':' . getenv('PORT');
$filename = getenv('FILENAME_DOWNLOAD');
$file_path = $path . '/' . $filename;
$read_file = '';

if (file_exists($file_path)) {
	$ofile = fopen($file_path, "r");
	$read_file = fread($ofile, filesize($file_path));
	fclose($ofile);
	$status = 200;
	$status_message = 'OK';
	$dic_header = [
		'Connection' => getenv('CONNECTION'),
		'Content-Type' => 'application/octet-stream',
		'Content-Length' => strlen($read_file),
		'Content-Disposition' => 'attachment; filename=' . $filename,
	];
} else {
	$ofile = fopen("./page/error/404.html", "r");
	$read_file = fread($ofile, filesize('./page/error/404.html'));
	fclose($ofile);
	$status = 404;
	$status_message = 'Not Found';
	$dic_header = [
		'Connection' => getenv('CONNECTION'),
		'Content-Type' => 'text/html',
		'Content-Length' => strlen($read_file),
	];
}

printHeaderBody($dic_header, $read_file, $status, $status_message);
?>

File: cgi-bin/python/download.py
#!/usr/bin/env python3

import os
import utils

try:
	path = utils.getEnvValue("ROOT_PATH") + ":" + utils.getEnvValue("PORT")
	filename = utils.getEnvValue("FILENAME_DOWNLOAD")
	file_path = os.path.join(path, filename)
	read_file = ""
	
	if (os.path.isfile(file_path)):
		with open(file_path, "r") as file:
			read_file = file.read()
		file.close()
	else:
		raise utils.InternalServerError
	dic_header = {
		"Connection" : utils.getEnvValue("CONNECTION"),
		"Content-Type" : "application/octet-stream",
		"Content-Length" : len(read_file) + 1,
		"Content-Disposition" : f'attachment; filename="{filename}"'
	}
	utils.printHeaderBody(dic_header, read_file, "200", "ok")
except (utils.InternalServerError,  FileNotFoundError):
	with open("./page/error/404.html", "r") as file:
		err_page = file.read()
	dic_header = {
		"Connection" : utils.getEnvValue("CONNECTION"),
		"Content-Type" : "text/html",
		"Content-Length" : len(err_page) + 1,
	}
	utils.printHeaderBody(dic_header, err_page, "404", "Not Found")

